crates/bevy_ecs/src/component.rs

```diff
@@ -261,7 +279,70 @@ impl ComponentInfo {
 
     /// Create a new [`ComponentInfo`].
     pub(crate) fn new(id: ComponentId, descriptor: ComponentDescriptor) -> Self {
-        ComponentInfo { id, descriptor }
+        ComponentInfo {
+            id,
+            descriptor,
+            hooks: ComponentHooks::default(),
+        }
+    }
+
+    /// Register a [`ComponentHook`] that will be run when this component is added to an entity
+    ///
+    /// Will panic if the component already has an `on_add` hook
+    pub fn on_add(&mut self, hook: ComponentHook) -> &mut Self {
```

Comment:
We should have fallible versions of these functions. If a plugin wants to add a hook to a component defined elsewhere, there should be some way of recovering.

Reply:
I couldn't think of many cases that they would be useful, perhaps if you can fallback to some other marker component instead. Happy to add them though.

Reply:
I agree that they probably won't be very useful in general, but we should at least provide plugin authors the option of avoiding a crash.